Refactor dashboard navlets to use HTMX

Simrayz opened the pull request:
## Scope and purpose

Resolves #3635.

This PR refactors the dashboard navlets logic from a complicated JS flow, to use HTMX where possible, and simplifies necessary javascript plugin interoperability.

See issue #3635 for a full description of changes.

<!-- remove things that do not apply -->
### This pull request
* Implements loading and updating of navlets with HTMX
* Replaces `navlet_controller` and `navlets_controller` with `navlets_htmx_controller`
* Renders entire dashboard on initial load of the page, with individual data fetching for each navlet
* Simplifies refresh of image- and ajax reload navlets

## How to test

### Loading navlets

#### Navlet Load

1. Open your default dashboard
2. Verify that navlets have a Spinner that indicates their loading state
3. Verify that all navlets load and resolve in their "VIEW" mode

#### Adding and Removing Navlets

1. Create a new dashboard. Verify that the dashboard has a "No widgets"
2. Add a new Navlet to your dashboard. 
   1. Verify that the Navlet is rendered in the leftmost column at the top, and that the "No widgets" message is removed.
   2. Verify that the navlet is highlighted, and that the highlight is removed on hover.
   3. Add another Navlet to your dashboard.
   3. Reload the page, and verify that positions have been saved.
3. Delete a Navlet. 
   1. Verify that it is removed from the dashboard.
   2. Remove all remaining navlets, and verify that a "No widgets" message is added to the left-most column.   

### Dashboard Layout

#### Re-ordering Navlets

1. Open a dashboard with many navlets, e.g. your Default dashboard.
2. Hold the drag handle, and move a navlet into a different column.
3. Reload the page, and verify that the positions were saved.

#### Updating columns

1. Open a dashboard with many navlets, e.g. your Default dashboard.
2. Open dashboard settings, and change the column count.
3. Verify that the dashboard is reloaded, and the column count is changed.

#### Compact Dashboard Mode

1. Open a dashboard with many navlets, e.g. your Default dashboard.
2. Change the column density from "Normal" to "Compact"
3. Verify that the spacing between navlets is removed.

### Editing Navlets
> Fill in section about how to test editing of navlets.

### Periodic Refresh of Navlets

#### Refresh with standard interval
This can be tested with all Navlets that have a `refresh_interval` defined, and don't have `ajax_reload` or `image_reload` set to `True`. I have selected RoomStatus as an example as the refresh interval is short.

1. Add a RoomStatus ("Rooms with active alerts") navlet to a dashboard.
2. Take note of the "Last update" time, and wait 30 seconds.
3. Verify that the Navlet is reloaded. You can also view this in the Network tab in dev tools.

#### Refresh with `ajax_reload=True`

* Create an empty dashboard.
* Add an `Alert`, `PDU` and `UPS` navlet.
* Configure the navlets to show content.
* The default refresh interval is 30 seconds, so open the network tab and look for repeating `render` requests.
* Verify that the "Last update" text in the bottom-right is updated every thirty seconds. Does not work for the Alert widget, if there is no data available in  graphite.

#### Refresh for VlanGraphNavlet and GraphWidget

1. Add a Vlan Graph to a dashboard.
   1. Go into edit mode, and select a Vlan. Not all of them have an URL, so try one that works, e.g.  VLAN 20 for employees.
   2. Verify that an image (or placeholder with No Data) is rendered.
2. Add a Chart (GraphWidget) to a dashboard.
   1. Go into edit mode. You need a Chart URL and a Target URL. I found the most practical way to find values for this form was to find a GraphWidget in the account_navlet table, and copy the values from preferences.
   2. Verify that an image (or placeholder with No Data) is rendered.

## Contributor Checklist

Every pull request should have this checklist filled out, no matter how small it is.
More information about contributing to NAV can be found in the
[Hacker's guide to NAV](https://nav.readthedocs.io/en/latest/hacking/hacking.html#hacker-s-guide-to-nav).

<!-- Add an "X" inside the brackets to confirm -->
<!-- If not checking one or more of the boxes, please explain why below each or remove the line if not applicable. -->

* [x] Added a changelog fragment for [towncrier](https://nav.readthedocs.io/en/latest/hacking/hacking.html#adding-a-changelog-entry)
* [x] Added/amended tests for new/changed code
* [ ] Added/changed documentation
* [x] Linted/formatted the code with ruff, easiest by using [pre-commit](https://nav.readthedocs.io/en/latest/hacking/hacking.html#pre-commit-hooks-and-ruff)
* [x] Wrote the commit message so that the first line continues the sentence "If applied, this commit will ...", starts with a capital letter, does not end with punctuation and is 50 characters or less long. See https://cbea.ms/git-commit/
* [x] Based this pull request on the correct upstream branch: For a patch/bugfix affecting the latest stable version, it should be based on that version's branch (`<major>.<minor>.x`). For a new feature or other additions, it should be based on `master`.
* [ ] If applicable: Created new issues if this PR does not fix the issue completely/there is further work to be done
* [x] If it's not obvious from a linked issue, described how to interact with NAV in order for a reviewer to observe the effects of this change first-hand (commands, URLs, UI interactions)
* [ ] If this results in changes in the UI: Added screenshots of the before and after
* [ ] If this adds a new Python source code file: Added the [boilerplate header](https://nav.readthedocs.io/en/latest/hacking/hacking.html#python-boilerplate-headers) to that file

<!-- Make this a draft PR if the content is subject to change, cannot be merged or if it is for initial feedback -->

Commit:
Remove old navlet controllers

Commit:
Add initial support for rendering dashboard with htmx

Commit:
Add support for adding and removing navlets with htmx

Commit:
Change column chooser to update with htmx

Commit:
Move navlet listeners to htmx controller

Commit:
Add support for saving navlet forms with HTMX

Commit:
Handle form errors in status navlet

Commit:
Support re-rendering of form select elements with htmx

Commit:
Add bust parameter to graph urls

Commit:
Add refresh template for handling periodic refresh

Commit:
Refactor navlets htmx controller

Commit:
Refactor navlet handlers in separate plugin

Commit:
Add changelog entry

Commit:
Support ajax reload with htmx swap

Commit:
Add dashboard loading tests

Commit:
Add navlet tests

Commit:
Add loading spinner to navlets

hmpf commented on python/nav/web/navlets/__init__.py:
```diff
@@ -142,7 +146,7 @@ def get_context_data_edit(self, context):
         return context
 
-    def post(self, _request, **kwargs):
+    def post(self, request, **kwargs):
```
Whoa, many old sins in this code, Liskov-wise...

hmpf commented on python/nav/web/navlets/graph.py:
```diff
@@ -56,7 +57,9 @@ def get_context_data_view(self, context):
-        context['graph_url'] = self.preferences.get('url')
+        url = self.preferences.get('url')
+        timestamp = int(datetime.now().timestamp())
+        context['graph_url'] = f'{url}&bust={timestamp}'
```
Cache-busting? Add a comment about it.

Simrayz replied:
The bust parameter was defined here. 
https://github.com/Uninett/nav/blob/16d641956c62877edc53e0a28eabe07842c5a871/python/nav/web/static/js/src/plugins/navlet_controller.js#L129-L139

I created a new staticmethod on the Navlet class, `add_bust_param_to_url`

hmpf commented on python/nav/web/templates/navlets/_add_navlet_modal.html:
```diff
@@ -3,7 +3,11 @@ <h4>Add a widget to your page</h4>
+                  class="add-user-navlet right"
+                  hx-post="{% url 'add-user-navlet' dashboard_id %}"
+                  hx-swap="beforeend"
+          >
```
What happens if you keep `action` and `post` in addition to what's here now?

lunkwill42 replied:
Is progressive fallback not a goal here? (not that I'm sure this ever had functional progressive fallback).

hmpf reviewed (changes requested):
Add one commit that fixes the get's and post's of all the TemplateView desecendants to also have `*args` (or `*_`) and `**kwargs` (or `**__`) even if they don use them. Devs used to Django expect it.

Simrayz commented on python/nav/web/navlets/vlangraph.py:
```diff
@@ -46,7 +47,8 @@ def get_context_data(self, **kwargs):
             url = get_vlan_graph_url(vlanid)
-            context['graph_url'] = url
+            timestamp = int(datetime.now().timestamp())
+            context['graph_url'] = f'{url}&bust={timestamp}'
```
I created a method on the Navlet class

johannaengland commented on python/nav/web/navlets/__init__.py:
```diff
@@ -266,7 +271,12 @@ def dispatcher(request, navlet_id):
+                request, account_navlet, 'Not authorized to view this widget'
+            )
+        else:
+            return HttpResponse(status=403)
```
Above the structure 
```
if request.htmx:
    return ...
return HttpResponse(...)
```
is used and here
```
if request.htmx:
    return ...
else:
    return HttpResponse(...)
```

I am happy with either variation, but it should be consistent

johannaengland commented on python/nav/web/webfront/views.py:
```diff
@@ -95,11 +99,45 @@ def index(request, did=None):
+    '2': 'medium-6',
+    '3': 'medium-4',
+    '4': 'medium-3',
+}
```
Please move the constant to the top of the file

johannaengland reviewed (changes requested):
When I set the column number to one I get a different behavior
Before:
<img width="1862" height="524" alt="image" src="https://github.com/user-attachments/assets/ac783626-d06b-4f67-9eb3-b64f39c230c8" />

After:
<img width="1862" height="524" alt="image" src="https://github.com/user-attachments/assets/0e5b1a4e-1a65-4982-b19f-676621569393" />

(I double checked that for both the column number is set to one)

When I try to add a "Rooms with active alerts" or "Locations with active alerts" widget this is all I get:
<img width="922" height="359" alt="image" src="https://github.com/user-attachments/assets/62b33b39-d2aa-49e3-85a1-b06b6e6fa644" />

Simrayz commented on python/nav/web/webfront/views.py:
```diff
@@ -632,10 +632,25 @@ def save_dashboard_columns(request, did):
+        "webfront/_dashboard_settings_columns_form.html",
+        {
+            'dashboard': dashboard,
+            'message': 'Dashboard columns updated to {}.'.format(new_column_count),
```
Done

johannaengland replied:
But now the tests also need to be adapted for that, because they are currently failing

Simrayz commented on tests/integration/web/navlets_test.py:
```diff
@@ -1,17 +1,18 @@
 
 from nav.models.profiles import Account, AccountDashboard, AccountNavlet
-from nav.web.navlets import add_navlet, modify_navlet
+from nav.web.navlets import add_navlet, get_navlet_from_name, modify_navlet, Navlet
```
Done

Simrayz commented on python/nav/web/webfront/views.py:
```diff
@@ -95,11 +99,45 @@ def index(request, did=None):
+
+def load_dashboard(request, dashboard_id=None):
+    """Renders the dashboard widgets for a given dashboard."""
+    dashboard = find_dashboard(request.account, dashboard_id)
```
You're right, this PR has been open so long that the auth changes were put into effect in the meantime :p

Simrayz commented:
@johannaengland Thanks for finding that bug! The problem was a missing css class when the column count is 1 🙃 It should now behave as normal:
<img width="1068" height="497" alt="image" src="https://github.com/user-attachments/assets/35a5ebe4-455a-482e-829a-a34c1b1d112f" />

johannaengland reviewed (changes requested):
The tests are failing due to the change in the column change message

I also still cannot get the "Rooms/Locations with active alerts" widget to work 

And I found a fun new bug, when logging in as a non-admin user and creating a new dashboard, I get the following view:

<img width="1825" height="939" alt="image" src="https://github.com/user-attachments/assets/d35a3ca1-d8d8-4cef-9f9b-e2007f91b6dd" />

Commit:
Fix authentication error for unauthenticated users

Simrayz commented:
@johannaengland I fixed the authentication error by adding the dashboard loading URL to the authentication middleware whitelist. I also fixed the failing test. 

The Room/Location Active Alert navlets seem to work for me.

<img width="719" height="320" alt="image" src="https://github.com/user-attachments/assets/7452f843-d74f-4817-a084-eea1013d4459" />

Commit:
Refactor navlets htmx controller and fix initialization bug

lunkwill42 commented on tests/integration/widget_test.py:
```diff
@@ -130,6 +130,44 @@ def test_given_navlet_id_when_navlet_type_is_invalid_then_return_error_widget(
     assert navlet.title == "Error"
 
 
+def test_given_htmx_request_with_non_existent_navlet_id_then_return_htmx_error_response(
```
```suggestion
def test_given_htmx_request_with_non_existent_navlet_id_then_it_should_return_htmx_error_response(
```

lunkwill42 reviewed (changes requested):
I'm not deeply familiar with how the dashboard/navlet system works to begin with, but this certainly does look like a massive improvement over the existing code.  Therefore, I don't have terribly much to say about the code itself (apart for my usual test naming nitpicks :laughing:), but I have tested in my browser (Firefox 145) and found a couple of weird things:

- Dashboard compact mode looks different than before, and has issues. Something funny happens with the internal padding of widgets, and there are gaps in the vertical space between widgets where there before were none.

- I added widgets of type "Rooms with active alerts" or "Locations with active alerts", then expanded (accordion?) one of the rooms or locations with active alerts to see the list of alerts under each. As soon as the widget is refreshed (on the refresh interval), my expansion is immediately forgotten and I am returned to the default state of the widget.  This does not appear to happen in the original version, and this definitely reduces usability. I have no idea how this was achieved in the original.

- There are non-fixup commits that exist only to refactor code that was added previously in the same PR.  While refactoring is fine, doing it within a single PR just makes it harder to review, so I would suggest squashing things into more logical commits when you get to the "squashing fixups"-stage.

- There are some comments about issues with rendering for unauthenticated users. I haven't the time or the brainpower to grok the issue this late in the afternoon, I'll have to get back to it :laughing:

Commit:
Fix bug where active accordion tab is not preserved when refreshing

Simrayz commented on python/nav/web/navlets/__init__.py:
```diff
@@ -116,21 +119,31 @@ def get_template_basename(self):
+        :param override_mode: Optional\; if provided, overrides the mode (VIEW or EDIT)
+            sent in the request. If None, uses self.mode. Used to enable the navlet to
+            return the correct template in error situations.
+        :type override_mode: str or None
```
Fixed :)

Commit:
Fix bug where navlets take up more vertical space than needed

lunkwill42 reviewed (changes requested):
> Try generating the css again and do a hard refresh, there is no bottom margin on navlets in compact mode :) I also fixed the height of navlets, I originally added a min-height to make the loading icon work as intended. It is now moved to a separate `.loading` class.

Well, no longer reproducible for me now so a 👍  from me there 😃 

> This is now fixed. The main issue was that jQuery UI sends the value as an integer, while html attributes always use strings... :') Why it worked before I'm unsure, but it works now and that's good enough for me.
> Screen.Recording.2025-12-11.at.21.21.35.mov

Looks much better!

New random manual test that fails for me: I added a PDU load widget. This widget needs to be configured before it can show anything. When empty, it shows a hyperlink to its own configuration page, and this link no longer seems to work after the refactor (i.e. it does nothing in my browser (except a slight hint: The word "clicked" is output on the console)

johannaengland reviewed (changes requested):
See my comment above, please also fix the linting and formatting errors

johannaengland reviewed (changes requested):
Fun new bug I just found, please have a test and see if you can reproduce it:

When I add a completely new dashboard and add at least two widgets I am not able to move the widgets. But if reload the page then it works

johannaengland reviewed (approved):
https://media2.giphy.com/media/v1.Y2lkPTc5MGI3NjExYTUxM3FobHA4ZHlibXJjNGEyc3B6eHZ4dzFucjJuZDFvNDA4NDN5YyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/whhCAcuxtAXcrQhHTk/giphy.gif
